pkg/service/providerindex/publisher/publisher.go

```diff
@@ -0,0 +1,190 @@
+package publisher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+
+	cid "github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipni/go-libipni/metadata"
+	provider "github.com/ipni/index-provider"
+	"github.com/ipni/index-provider/engine"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	mh "github.com/multiformats/go-multihash"
+)
+
+var log = logging.Logger("publisher")
+var remoteHeadPrefix = datastore.NewKey("head/remote/")
+
+type Publisher interface {
+	// Publish publishes an advert to indexer(s). Note: it is not necessary to
+	// sign the advert - this is done automatically.
+	Publish(ctx context.Context, provider *peer.AddrInfo, contextID string, digests []mh.Multihash, meta metadata.Metadata) (ipld.Link, error)
+	// NotifyRemoteSync calls the notification function when the configured remote
+	// IPNI node is observed to have performed a sync.
+	NotifyRemoteSync(NotifyRemoteSyncFunc)
+	// Store returns the storage interface used to access published data.
+	Store() AdvertStore
+}
+
+type IPNIPublisher struct {
+	engine   *engine.Engine
+	notifier *Notifier
+	store    AdvertStore
+	mutex    sync.RWMutex
+}
+
+func (p *IPNIPublisher) NotifyRemoteSync(f NotifyRemoteSyncFunc) {
+	if p.notifier != nil {
+		p.notifier.Notify(f)
+	} else {
+		log.Errorf("notification for sync events requested but no remote IPNI node configured")
+	}
+}
+
+func (p *IPNIPublisher) Start(ctx context.Context) error {
+	err := p.engine.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("starting IPNI index provider: %w", err)
+	}
+	if p.notifier != nil {
+		p.notifier.Start(ctx)
+	}
+	return nil
+}
+
+func (p *IPNIPublisher) Publish(ctx context.Context, providerInfo *peer.AddrInfo, contextID string, digests []mh.Multihash, meta metadata.Metadata) (ipld.Link, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	// such a weird API
+	p.engine.RegisterMultihashLister(func(ctx context.Context, _ peer.ID, _ []byte) (provider.MultihashIterator, error) {
```

Comment:
Ok, so here's why this is a weird API. We need to call "publish" (`NotifyPut`) sequentially because each new advert needs to link back to the current latest. There's no lock in the IPNI `index-provider` code that ensures my advert doesn't get overwritten by another if trying to publish 2 at the same time. This is why the lock is here.

So since it's sequential there's only 1 process reading from the lister at any given moment, so if we lock, then _set_ the lister and then publish, we don't need to check the context ID, because what is being requested is what we're publishing.

IDK I'm having a hard time interfacing with this code so perhaps there's something here I'm missing? I don't understand why you wouldn't pass a slice (or iterable) to `NotifyPut` directly. Can you help? 🙏

Reply:
If I remember correctly (and it's been a minute), the idea is you have just one multihash lister who you register long before you ever call NotifyPut, and then for a context ID, you go and grab the multihashes in the moment when the Lister is called.

And the reason for this would be in the original Filecoin SP case, they're going to want to use the context ID to go lookup a piece somewhere, and then return a multihash lister that iterates through the car file in the piece.

I still agree it would be simpler to just provider a MultihashLister to notify put.

The whole index-provider code is very opinionated and tailored for a specific type of operation, which is why I suggested trying to just use the primitives in go-libipni to start with. But the problem is the index provider does do a lot and rebuilding from the primitives would be laborious. Maybe we can make a TODO at somepoint to copy paste only the relevant parts of index-provider code we really need and then just build on the go-libipni primitives. But that's a tomorrow problem :)

Reply:
> But the problem is the index provider does do a lot and rebuilding from the primitives would be laborious.

Ya, exactly what I ran into.

> Maybe we can make a TODO at somepoint to copy paste only the relevant parts of index-provider code we really need and then just build on the go-libipni primitives. But that's a tomorrow problem :)

https://github.com/storacha/indexing-service/issues/19